src/main/java/ch/njol/skript/sections/SecFilter.java

```diff
@@ -0,0 +1,191 @@
+package ch.njol.skript.sections;
+
+import ch.njol.skript.ScriptLoader;
+import ch.njol.skript.Skript;
+import ch.njol.skript.config.Node;
+import ch.njol.skript.config.SectionNode;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.ExprInput;
+import ch.njol.skript.lang.Condition;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.InputSource;
+import ch.njol.skript.lang.Section;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.TriggerItem;
+import ch.njol.skript.lang.Variable;
+import ch.njol.skript.lang.parser.ParserInstance;
+import ch.njol.skript.variables.Variables;
+import ch.njol.util.Kleenean;
+import ch.njol.util.Pair;
+import ch.njol.util.StringUtils;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.UnknownNullability;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.StreamSupport;
+
+@Name("Filter (Section)")
+@Description({
+	"Filters a variable list based on the supplied conditions. Unlike the filter expression, this effect " +
+	"maintains the indices of the filtered list.",
+	"It also supports filtering based on meeting any of the given criteria, rather than all, like multi-line if statements."
+})
+@Examples({
+	"set {_a::*} to integers between -10 and 10",
+	"",
+	"filter {_a::*} to match:",
+		"\tinput is a number",
+		"\tmod(input, 2) = 0",
+		"\tinput > 0",
+	"",
+	"send {_a::*} # sends 2, 4, 6, 8, and 10",
+})
+@Since("INSERT VERSION")
+public class SecFilter extends Section implements InputSource {
+
+	static {
+		Skript.registerSection(SecFilter.class,
+				"filter %objects% to match [:any|all]");
+		if (!ParserInstance.isRegistered(InputSource.InputData.class))
+			ParserInstance.registerData(InputSource.InputData.class, InputSource.InputData::new);
+	}
+
+	@UnknownNullability
+	private Variable<?> unfilteredObjects;
+	private final List<Condition> conditions = new ArrayList<>();
+	private boolean isAny;
+
+	@Nullable
+	private Object currentValue;
+	@UnknownNullability
+	private String currentIndex;
```

Comment:
move these onto the type? 🙂

Reply:
I will do this shortly in a following PR! some refactoring is coming to all these input classes in general